File: lc-typescript/src/lc_0733_flood_fill.ts
import { assertEquals } from "https://deno.land/std@0.224.0/assert/assert_equals.ts";

function flood_fill1(
    image: number[][],
    sr: number,
    sc: number,
    color: number,
): number[][] {
    const starting_color = image[sr][sc];

    const stack: [number, number][] = [[sr, sc]];
    while (stack.length > 0) {
        const [row, col] = stack.pop()!;
        if (image[row][col] !== starting_color || image[row][col] === color) {
            continue;
        }

        image[row][col] = color;

        if (row - 1 >= 0) {
            stack.push([row - 1, col]);
        }
        if (row + 1 < image.length) {
            stack.push([row + 1, col]);
        }
        if (col - 1 >= 0) {
            stack.push([row, col - 1]);
        }
        if (col + 1 < image[sr].length) {
            stack.push([row, col + 1]);
        }
    }

    return image;
}

function flood_fill2(
    image: number[][],
    sr: number,
    sc: number,
    color: number,
): number[][] {
    const scolor = image[sr][sc];
    if (scolor === color) {
        return image;
    }

    const stack = [[sr, sc]];
    while (stack.length > 0) {
        const [r, c] = stack.pop()!;
        image[r][c] = color;
        if (r + 1 < image.length && image[r + 1][c] === scolor) {
            stack.push([r + 1, c]);
        }
        if (c + 1 < image[r].length && image[r][c + 1] === scolor) {
            stack.push([r, c + 1]);
        }
        if (r - 1 >= 0 && image[r - 1][c] === scolor) {
            stack.push([r - 1, c]);
        }
        if (c - 1 >= 0 && image[r][c - 1] === scolor) {
            stack.push([r, c - 1]);
        }
    }

    return image;
}

Deno.test("0733 Flood Fill", () => {
    function get_test_cases() {
        const test_cases: [number[][], number, number, number, number[][]][] = [
            [
                [[1, 1, 1], [1, 1, 0], [1, 0, 1]],
                1,
                1,
                2,
                [[2, 2, 2], [2, 2, 0], [2, 0, 1]],
            ],
            [
                [[0, 0, 0], [0, 0, 0]],
                0,
                0,
                0,
                [[0, 0, 0], [0, 0, 0]],
            ],
        ];
        return test_cases;
    }

    const ts1 = get_test_cases();
    for (let i = 0; i < ts1.length; i += 1) {
        const test_case = ts1[i];

        const image = test_case[0];
        const sr = test_case[1];
        const sc = test_case[2];
        const color = test_case[3];

        const expected = test_case[4];
        const actual = flood_fill1(image, sr, sc, color);

        assertEquals(actual, expected);
    }

    const ts2 = get_test_cases();
    for (let i = 0; i < ts2.length; i += 1) {
        const test_case = ts2[i];

        const image = test_case[0];
        const sr = test_case[1];
        const sc = test_case[2];
        const color = test_case[3];

        const expected = test_case[4];
        const actual = flood_fill2(image, sr, sc, color);

        assertEquals(actual, expected);
    }
});

File: lc-python/src/solutions/lc_0733_flood_fill.py
import unittest
from typing import List


def flood_fill0(
    image: List[List[int]], row: int, col: int, color: int
) -> List[List[int]]:
    """
    Time complexity: O(n*m)
    Space complexity: O(n*m)
    """
    # https://en.wikipedia.org/wiki/Flood_fill#Stack-based_recursive_implementation_(four-way)
    color_old = image[row][col]

    # depth-first graph traversal
    to_fill = [(row, col)]
    while len(to_fill) > 0:
        x, y = to_fill.pop()

        if image[x][y] != color_old or image[x][y] == color:
            continue

        image[x][y] = color

        # next valid pixel
        if x > 0:
            to_fill.append((x - 1, y))
        if x < len(image) - 1:
            to_fill.append((x + 1, y))
        if y > 0:
            to_fill.append((x, y - 1))
        if y < len(image[x]) - 1:
            to_fill.append((x, y + 1))

    return image


def flood_fill1(
    image: List[List[int]], row: int, col: int, color: int
) -> List[List[int]]:
    def dfs(x: int, y: int) -> None:
        if image[x][y] != color_old or image[x][y] == color:
            return

        image[x][y] = color

        # next valid pixel
        if x > 0:
            dfs(x - 1, y)
        if x < len(image) - 1:
            dfs(x + 1, y)
        if y > 0:
            dfs(x, y - 1)
        if y < len(image[x]) - 1:
            dfs(x, y + 1)

    color_old = image[row][col]

    dfs(row, col)

    return image


class TestFloodFill(unittest.TestCase):

    @staticmethod
    def parse_input():
        data = [
            [
                [[1, 1, 1], [1, 1, 0], [1, 0, 1]],
                1,
                1,
                2,
                [[2, 2, 2], [2, 2, 0], [2, 0, 1]],
            ],
            [[[0, 0, 0], [0, 0, 0]], 0, 0, 0, [[0, 0, 0], [0, 0, 0]]],
        ]
        return data

    def test_flood_fill0(self):
        test_cases = self.parse_input()
        for i, tc in enumerate(test_cases):
            actual = flood_fill0(tc[0], tc[1], tc[2], tc[3])
            expected = tc[4]
            self.assertEqual(actual, expected, f"{i} {tc}")

    def test_flood_fill1(self):
        test_cases = self.parse_input()
        for i, tc in enumerate(test_cases):
            actual = flood_fill1(tc[0], tc[1], tc[2], tc[3])
            expected = tc[4]
            self.assertEqual(actual, expected, f"{i} {tc}")


if __name__ == "__main__":
    unittest.main()
